[TS migration] Improve ref types (#40698)
Expensify/App: contributingGuides/TS_STYLE.md

  export someVariable
  ```

> - [1.23](#ref-types) **Ref types**: Avoid using HTML elements while declaring refs. Please use React Native components where possible. React Native Web handles the references on its own. It also extends React Native components [with Interaction API](https://necolas.github.io/react-native-web/docs/interactions/) which should be used to handle Pointers and Mouse events. If methods  Exception of this rule is when we explicitly need to use functions available only in DOM and not in React Native, e.g. `getBoundingClientRect`. Then please declare ref type as `union` of React Native component and HTML element. When passing it to React Native component cast it as soon as possible using utility methods declared in `src/types/utils`

Reviewer: ```suggestion
- [1.23](#ref-types) **Ref types**: Avoid using HTML elements while declaring refs. Please use React Native components where possible. React Native Web handles the references on its own. It also extends React Native components [with Interaction API](https://necolas.github.io/react-native-web/docs/interactions/) which should be used to handle Pointers and Mouse events. Exception of this rule is when we explicitly need to use functions available only in DOM and not in React Native, e.g. `getBoundingClientRect`. Then please declare ref type as `union` of React Native component and HTML element. When passing it to React Native component cast it as soon as possible using utility methods declared in `src/types/utils`
```

```ts
import viewRef from '@src/types/utils/viewRef';

const ref = useRef<View | HTMLDivElement>();

if (ref.current && 'getBoundingClientRect' in ref.current ){
  ref.current.getBoundingClientRect();
}

<View ref={viewRef(ref)}>

> ```

Reviewer: Let's add an example how it should be done usually with just `View` type, and the example above should be marked as an exception (we could split the above text into two to better visualize the meaning)

  export someVariable
  ```

> - [1.23](#ref-types) **Ref types**: Avoid using HTML elements while declaring refs. Please use React Native components where possible. React Native Web handles the references on its own. It also extends React Native components [with Interaction API](https://necolas.github.io/react-native-web/docs/interactions/) which should be used to handle Pointers and Mouse events. Exception of this rule is when we explicitly need to use functions available only in DOM and not in React Native, e.g. `getBoundingClientRect`. Then please declare ref type as `union` of React Native component and HTML element. When passing it to React Native component cast it as soon as possible using utility methods declared in `src/types/utils`

Reviewer 2: ```suggestion
- [1.23](#ref-types) **Ref types**: Avoid using HTML elements while declaring refs. Please use React Native components where possible. React Native Web handles the references on its own. It also extends React Native components with [Interaction API](https://necolas.github.io/react-native-web/docs/interactions/) which should be used to handle Pointer and Mouse events. Exception of this rule is when we explicitly need to use functions available only in DOM and not in React Native, e.g. `getBoundingClientRect`. Then please declare ref type as `union` of React Native component and HTML element. When passing it to React Native component assert it as soon as possible using utility methods declared in `src/types/utils`.
```

nits

Normal usage: 

> ```ts

Reviewer 2: ```suggestion
```tsx
```

const ref = useRef<View>();

<View ref={ref} onPointerDown={e => {#DO SOMETHING}}>
```

Exceptional usage where DOM methods are necessary:

> ```ts

Reviewer 2: ```suggestion
```tsx
```

import viewRef from '@src/types/utils/viewRef';

const ref = useRef<View | HTMLDivElement>();

> if (ref.current && 'getBoundingClientRect' in ref.current ){

Reviewer 2: ```suggestion
if (ref.current && 'getBoundingClientRect' in ref.current) {
```

> if (ref.current && 'getBoundingClientRect' in ref.current ) {

Reviewer 2: ```suggestion
if (ref.current && 'getBoundingClientRect' in ref.current) {
```
